src/frontend/Ast_to_Mir.ml

```diff
@@ -451,14 +457,19 @@ let rec trans_stmt declc (ts : Ast.typed_statement) =
   | Ast.While (cond, body) ->
       While (trans_expr cond, trans_single_stmt body) |> swrap
   | Ast.For {loop_variable; lower_bound; upper_bound; loop_body} ->
+      let body =
+        match trans_single_stmt loop_body with
+        | {stmt= Block _; _} as b -> b
+        | x -> {x with stmt= Block [x]}
+      in
       For
         { loopvar= loop_variable.Ast.name
         ; lower= trans_expr lower_bound
         ; upper= trans_expr upper_bound
-        ; body= trans_single_stmt loop_body }
+        ; body }
       |> swrap
   | Ast.ForEach (loopvar, iteratee, body) ->
-      let newsym = Util.gensym () in
+      let newsym = gensym () in
       let wrap expr = {expr; emeta= {mloc; mtype= UInt; madlevel= DataOnly}} in
```

Comment:
I think the translation of the ForEach loop is still wrong.
It should probably first generate a fresh loop variable like you do.
Then it should inspect the type of the iteratee and distinguish between Eigen types and Arrays in the translation. It should calculate the correct bounds. And start the loop body with the declaration and definition of the old loop variable of the Foreach loop in terms of the new loop variable and the iteratee.
I think @bob-carpenter may have already written out what the code gen should look like.